File: ScaleformUI_Csharp/Menus/UIMenu/Panels/UIMenuGridPanel.cs
using CitizenFX.Core;
using CitizenFX.Core.Native;
using System.Drawing;

namespace ScaleformUI.Menu
{
    public enum GridType
    {
        Full,
        Horizontal
    }

    public class UIMenuGridPanel : UIMenuPanel
    {
        public string TopLabel { get; set; }
        public string LeftLabel { get; set; }
        public string RightLabel { get; set; }
        public string BottomLabel { get; set; }
        public event GridPanelChangedEvent OnGridPanelChange;
        public GridType GridType = GridType.Full;

        private UIMenuGridAudio Audio;
        internal PointF _value;
        public PointF CirclePosition
        {
            get
            {
                return _value;
            }
            set
            {
                _value = value;
                if (ParentItem != null && ParentItem.Parent != null && ParentItem.Parent.Visible)
                {
                    int it = ParentItem.Parent.MenuItems.IndexOf(ParentItem);
                    ParentItem.Parent.SendPanelsToItemScaleform(it, true);
                    OnGridChange();
                    ParentItem.Parent.GridPanelChange(ParentItem, this, _value);
                }
            }
        }

        /// <summary>
        /// Creates <see langword="abstract"/>full grid panels with x,y values
        /// </summary>
        /// <param name="TopText"></param>
        /// <param name="LeftText"></param>
        /// <param name="RightText"></param>
        /// <param name="BottomText"></param>
        /// <param name="circlePosition"></param>
        public UIMenuGridPanel(string TopText, string LeftText, string RightText, string BottomText, PointF circlePosition)
        {
            TopLabel = TopText ?? "Up";
            LeftLabel = LeftText ?? "Left";
            RightLabel = RightText ?? "Right";
            BottomLabel = BottomText ?? "Down";
            GridType = GridType.Full;
            _value = circlePosition;
        }

        /// <summary>
        /// Creates an horizontal Panel with y fixed at 0.5 and variable x
        /// </summary>
        /// <param name="LeftText"></param>
        /// <param name="RightText"></param>
        /// <param name="circlePosition"></param>
        public UIMenuGridPanel(string LeftText, string RightText, PointF circlePosition)
        {
            TopLabel = "";
            BottomLabel = "";
            LeftLabel = LeftText ?? "Left";
            RightLabel = RightText ?? "Right";
            GridType = GridType.Horizontal;
            _value = circlePosition;
        }

        internal void OnGridChange()
        {
            OnGridPanelChange?.Invoke(ParentItem, this, CirclePosition);
        }
    }


    public class UIMenuGridAudio
    {
        public string Slider;
        public string Library;
        public int Id;
        public UIMenuGridAudio(string slider, string library, int id)
        {
            Slider = slider;
            Library = library;
            Id = id;
        }
    }
}

File: ScaleformUI_Lua/src/Menus/UIMenu/panels/UIMenuGridPanel.lua
UIMenuGridPanel = setmetatable({}, UIMenuGridPanel)
UIMenuGridPanel.__index = UIMenuGridPanel
UIMenuGridPanel.__call = function() return "UIMenuPanel", "UIMenuGridPanel" end

---@class UIMenuGridPanel
---@field public TopLabel string
---@field public RightLabel string
---@field public LeftLabel string
---@field public BottomLabel string
---@field public GridType number
---@field public ParentItem table
---@field public SetParentItem fun(self:UIMenuStatisticsPanel, item:UIMenuItem):UIMenuItem -- required
---@field public OnGridPanelChanged function

---New
---@param topText string
---@param leftText string
---@param rightText string
---@param bottomText string
---@param circlePosition vector2
---@param gridType number
function UIMenuGridPanel.New(topText, leftText, rightText, bottomText, circlePosition, gridType)
    local _UIMenuGridPanel = {
        TopLabel = topText or "UP",
        RightLabel = rightText or "RIGHT",
        LeftLabel = leftText or "LEFT",
        BottomLabel = bottomText or "DOWN",
        _CirclePosition = circlePosition or vector2(0.5, 0.5),
        GridType = gridType or 0,
        ParentItem = nil, -- required
        OnGridPanelChanged = function(item, panel, newindex)
        end
    }
    return setmetatable(_UIMenuGridPanel, UIMenuGridPanel)
end

---SetParentItem
---@param Item table
function UIMenuGridPanel:SetParentItem(Item) -- required
    if Item() ~= nil then
        self.ParentItem = Item
    else
        return self.ParentItem
    end
end

function UIMenuGridPanel:CirclePosition(position)
    if position ~= nil then
        self._CirclePosition = position
        if self.ParentItem ~= nil and self.ParentItem:SetParentMenu() ~= nil and self.ParentItem:SetParentMenu():Visible() then
            local it = IndexOf(self.ParentItem:SetParentMenu().Items, self.ParentItem)
            self.ParentItem:SetParentMenu():SendPanelsToItemScaleform(it, true)
            self.OnGridPanelChanged(self.ParentItem, self, self._CirclePosition)
            self.ParentItem:SetParentMenu().OnGridPanelChanged(self.ParentItem, self, self._CirclePosition)
        end
    else
        return self._CirclePosition
    end
end
